{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {},
   "outputs": [],
   "source": [
    "from random import seed, choice\n",
    "\n",
    "import os\n",
    "import heapq\n",
    "import numpy as np\n",
    "import scipy as sp\n",
    "import pandas as pd\n",
    "import matplotlib as mpl\n",
    "import matplotlib.pyplot as plt\n",
    "import seaborn as sns\n",
    "from skbio.tree import TreeNode"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Overview\n",
    "**Goal**: Generate a smaller group of clades from the original 10,575 genomes in the phylogeny, whereby the divergence amongst the clades can be limited to a given threshold, and the divergence between the clades is maximized.\n",
    "\n",
    "This approach is motivated by [WoL](https://github.com/biocore/wol/blob/master/code/notebooks/taxon_subsampling.ipynb) (Zhu et al.)\n",
    "\n",
    "First, we limit the possible groupings of clades to be chosen from those that have a minimum relative evolutionary divergence (**RED**) ([Parks, et al., 2018](https://www.nature.com/articles/nbt.4229)) over a given threshold, `min_red`.\n",
    "Then, for each candidate grouping rooted at `clade_root`, to maximize the divergence between clades, we can find the `n_clades` by finding the `n` descendants of `clade_root` that minimize $\\sum_{i=1}^n \\text{RED}(\\texttt{node}_i)$. Specific genomes can then be sampled based on criteria such as:\n",
    "1. Contains the most marker genes.\n",
    "2. Contamination level is the lowest.\n",
    "3. DNA quality score is the highest.\n",
    "4. Random selection.\n",
    "5. Sampling of all included genomes.\n",
    "\n",
    "Qiyun has previously used the first three in unison to select a single. Those three criteria could potentially be used to filter to a smaller list, amongst which all genomes are sampled.\n",
    "\n",
    "**Effect of paramters**:\n",
    "* Increasing `min_red` will limit the number of genomes a given group of clades can contain.\n",
    "* Increasing `n_clades` will increase the number of subgroups of genomes that are created.\n",
    "Increasing either paramter will increase the resolution that downstream methods need to properly handle such groups of genomes."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {},
   "outputs": [],
   "source": [
    "seed(42)\n",
    "%matplotlib inline"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "10575"
      ]
     },
     "execution_count": 17,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "tree_fp = 'https://raw.githubusercontent.com/biocore/wol/master/data/trees/astral/branch_length/cons/astral.cons.nid.nwk'\n",
    "# tree_fp = 'data/trees/astral.nid.nwk'\n",
    "tree = TreeNode.read(tree_fp)\n",
    "tree.count(tips=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "metadata": {},
   "outputs": [],
   "source": [
    "supports_fp = 'https://raw.githubusercontent.com/biocore/wol/master/data/trees/astral/astral.supports.tsv.bz2'\n",
    "dfs = pd.read_table(supports_fp, index_col=0)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>EN</th>\n",
       "      <th>LPP</th>\n",
       "      <th>QT</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>#node</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <td>N2</td>\n",
       "      <td>196.0</td>\n",
       "      <td>0.998406</td>\n",
       "      <td>0.450953</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>N3</td>\n",
       "      <td>196.0</td>\n",
       "      <td>0.998406</td>\n",
       "      <td>0.450953</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>N4</td>\n",
       "      <td>124.0</td>\n",
       "      <td>0.999993</td>\n",
       "      <td>0.535120</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>N5</td>\n",
       "      <td>195.0</td>\n",
       "      <td>0.914387</td>\n",
       "      <td>0.398648</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>N6</td>\n",
       "      <td>208.0</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.539983</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "          EN       LPP        QT\n",
       "#node                           \n",
       "N2     196.0  0.998406  0.450953\n",
       "N3     196.0  0.998406  0.450953\n",
       "N4     124.0  0.999993  0.535120\n",
       "N5     195.0  0.914387  0.398648\n",
       "N6     208.0  1.000000  0.539983"
      ]
     },
     "execution_count": 19,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "dfs.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "metadata": {},
   "outputs": [],
   "source": [
    "genomes_fp = 'https://biocore.github.io/wol/data/genomes/metadata.tsv.bz2'\n",
    "dfg = pd.read_table(genomes_fp, index_col=0)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "metadata": {
    "scrolled": false
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>asm_name</th>\n",
       "      <th>assembly_accession</th>\n",
       "      <th>bioproject</th>\n",
       "      <th>biosample</th>\n",
       "      <th>wgs_master</th>\n",
       "      <th>seq_rel_date</th>\n",
       "      <th>submitter</th>\n",
       "      <th>ftp_path</th>\n",
       "      <th>img_id</th>\n",
       "      <th>gtdb_id</th>\n",
       "      <th>...</th>\n",
       "      <th>coding_density</th>\n",
       "      <th>completeness</th>\n",
       "      <th>contamination</th>\n",
       "      <th>strain_heterogeneity</th>\n",
       "      <th>markers</th>\n",
       "      <th>5s_rrna</th>\n",
       "      <th>16s_rrna</th>\n",
       "      <th>23s_rrna</th>\n",
       "      <th>trnas</th>\n",
       "      <th>draft_quality</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>#genome</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <td>G000005825</td>\n",
       "      <td>ASM582v2</td>\n",
       "      <td>GCF_000005825.2</td>\n",
       "      <td>PRJNA224116</td>\n",
       "      <td>SAMN02603086</td>\n",
       "      <td>NaN</td>\n",
       "      <td>2010/12/15</td>\n",
       "      <td>Center for Genomic Sciences, Allegheny-Singer ...</td>\n",
       "      <td>ftp://ftp.ncbi.nlm.nih.gov/genomes/all/GCF/000...</td>\n",
       "      <td>646311908</td>\n",
       "      <td>RS_GCF_000005825.2</td>\n",
       "      <td>...</td>\n",
       "      <td>85.144124</td>\n",
       "      <td>98.68</td>\n",
       "      <td>1.32</td>\n",
       "      <td>0.0</td>\n",
       "      <td>377</td>\n",
       "      <td>yes</td>\n",
       "      <td>yes</td>\n",
       "      <td>yes</td>\n",
       "      <td>20</td>\n",
       "      <td>high</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>G000006175</td>\n",
       "      <td>ASM617v2</td>\n",
       "      <td>GCF_000006175.1</td>\n",
       "      <td>PRJNA224116</td>\n",
       "      <td>SAMN00000040</td>\n",
       "      <td>NaN</td>\n",
       "      <td>2010/06/03</td>\n",
       "      <td>US DOE Joint Genome Institute (JGI-PGF)</td>\n",
       "      <td>ftp://ftp.ncbi.nlm.nih.gov/genomes/all/GCF/000...</td>\n",
       "      <td>646564549</td>\n",
       "      <td>RS_GCF_000006175.1</td>\n",
       "      <td>...</td>\n",
       "      <td>80.167033</td>\n",
       "      <td>99.05</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.0</td>\n",
       "      <td>165</td>\n",
       "      <td>no</td>\n",
       "      <td>yes</td>\n",
       "      <td>yes</td>\n",
       "      <td>19</td>\n",
       "      <td>medium</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>G000006605</td>\n",
       "      <td>ASM660v1</td>\n",
       "      <td>GCF_000006605.1</td>\n",
       "      <td>PRJNA224116</td>\n",
       "      <td>SAMEA3283089</td>\n",
       "      <td>NaN</td>\n",
       "      <td>2005/06/27</td>\n",
       "      <td>Bielefeld Univ</td>\n",
       "      <td>ftp://ftp.ncbi.nlm.nih.gov/genomes/all/GCF/000...</td>\n",
       "      <td>637000085</td>\n",
       "      <td>RS_GCF_000006605.1</td>\n",
       "      <td>...</td>\n",
       "      <td>89.378688</td>\n",
       "      <td>100.00</td>\n",
       "      <td>0.68</td>\n",
       "      <td>0.0</td>\n",
       "      <td>319</td>\n",
       "      <td>yes</td>\n",
       "      <td>yes</td>\n",
       "      <td>yes</td>\n",
       "      <td>20</td>\n",
       "      <td>high</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>G000006725</td>\n",
       "      <td>ASM672v1</td>\n",
       "      <td>GCF_000006725.1</td>\n",
       "      <td>PRJNA224116</td>\n",
       "      <td>SAMN02603773</td>\n",
       "      <td>NaN</td>\n",
       "      <td>2004/06/04</td>\n",
       "      <td>Sao Paulo state (Brazil) Consortium</td>\n",
       "      <td>ftp://ftp.ncbi.nlm.nih.gov/genomes/all/GCF/000...</td>\n",
       "      <td>637000348</td>\n",
       "      <td>RS_GCF_000006725.1</td>\n",
       "      <td>...</td>\n",
       "      <td>82.592990</td>\n",
       "      <td>99.59</td>\n",
       "      <td>0.18</td>\n",
       "      <td>0.0</td>\n",
       "      <td>325</td>\n",
       "      <td>yes</td>\n",
       "      <td>yes</td>\n",
       "      <td>yes</td>\n",
       "      <td>20</td>\n",
       "      <td>high</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>G000006745</td>\n",
       "      <td>ASM674v1</td>\n",
       "      <td>GCF_000006745.1</td>\n",
       "      <td>PRJNA57623</td>\n",
       "      <td>SAMN02603969</td>\n",
       "      <td>NaN</td>\n",
       "      <td>2001/01/09</td>\n",
       "      <td>TIGR</td>\n",
       "      <td>ftp://ftp.ncbi.nlm.nih.gov/genomes/all/GCF/000...</td>\n",
       "      <td>637000333</td>\n",
       "      <td>RS_GCF_000006745.1</td>\n",
       "      <td>...</td>\n",
       "      <td>86.533164</td>\n",
       "      <td>99.86</td>\n",
       "      <td>0.03</td>\n",
       "      <td>0.0</td>\n",
       "      <td>360</td>\n",
       "      <td>yes</td>\n",
       "      <td>yes</td>\n",
       "      <td>yes</td>\n",
       "      <td>20</td>\n",
       "      <td>high</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows × 52 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "            asm_name assembly_accession   bioproject     biosample wgs_master  \\\n",
       "#genome                                                                         \n",
       "G000005825  ASM582v2    GCF_000005825.2  PRJNA224116  SAMN02603086        NaN   \n",
       "G000006175  ASM617v2    GCF_000006175.1  PRJNA224116  SAMN00000040        NaN   \n",
       "G000006605  ASM660v1    GCF_000006605.1  PRJNA224116  SAMEA3283089        NaN   \n",
       "G000006725  ASM672v1    GCF_000006725.1  PRJNA224116  SAMN02603773        NaN   \n",
       "G000006745  ASM674v1    GCF_000006745.1   PRJNA57623  SAMN02603969        NaN   \n",
       "\n",
       "           seq_rel_date                                          submitter  \\\n",
       "#genome                                                                      \n",
       "G000005825   2010/12/15  Center for Genomic Sciences, Allegheny-Singer ...   \n",
       "G000006175   2010/06/03            US DOE Joint Genome Institute (JGI-PGF)   \n",
       "G000006605   2005/06/27                                     Bielefeld Univ   \n",
       "G000006725   2004/06/04                Sao Paulo state (Brazil) Consortium   \n",
       "G000006745   2001/01/09                                               TIGR   \n",
       "\n",
       "                                                     ftp_path     img_id  \\\n",
       "#genome                                                                    \n",
       "G000005825  ftp://ftp.ncbi.nlm.nih.gov/genomes/all/GCF/000...  646311908   \n",
       "G000006175  ftp://ftp.ncbi.nlm.nih.gov/genomes/all/GCF/000...  646564549   \n",
       "G000006605  ftp://ftp.ncbi.nlm.nih.gov/genomes/all/GCF/000...  637000085   \n",
       "G000006725  ftp://ftp.ncbi.nlm.nih.gov/genomes/all/GCF/000...  637000348   \n",
       "G000006745  ftp://ftp.ncbi.nlm.nih.gov/genomes/all/GCF/000...  637000333   \n",
       "\n",
       "                       gtdb_id  ... coding_density completeness contamination  \\\n",
       "#genome                         ...                                             \n",
       "G000005825  RS_GCF_000005825.2  ...      85.144124        98.68          1.32   \n",
       "G000006175  RS_GCF_000006175.1  ...      80.167033        99.05          0.00   \n",
       "G000006605  RS_GCF_000006605.1  ...      89.378688       100.00          0.68   \n",
       "G000006725  RS_GCF_000006725.1  ...      82.592990        99.59          0.18   \n",
       "G000006745  RS_GCF_000006745.1  ...      86.533164        99.86          0.03   \n",
       "\n",
       "           strain_heterogeneity markers  5s_rrna  16s_rrna 23s_rrna trnas  \\\n",
       "#genome                                                                     \n",
       "G000005825                  0.0     377      yes       yes      yes    20   \n",
       "G000006175                  0.0     165       no       yes      yes    19   \n",
       "G000006605                  0.0     319      yes       yes      yes    20   \n",
       "G000006725                  0.0     325      yes       yes      yes    20   \n",
       "G000006745                  0.0     360      yes       yes      yes    20   \n",
       "\n",
       "           draft_quality  \n",
       "#genome                   \n",
       "G000005825          high  \n",
       "G000006175        medium  \n",
       "G000006605          high  \n",
       "G000006725          high  \n",
       "G000006745          high  \n",
       "\n",
       "[5 rows x 52 columns]"
      ]
     },
     "execution_count": 21,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "dfg.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "metadata": {},
   "outputs": [],
   "source": [
    "dfg.insert(0, '# assembly_accession', dfg['assembly_accession'])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th># assembly_accession</th>\n",
       "      <th>asm_name</th>\n",
       "      <th>assembly_accession</th>\n",
       "      <th>bioproject</th>\n",
       "      <th>biosample</th>\n",
       "      <th>wgs_master</th>\n",
       "      <th>seq_rel_date</th>\n",
       "      <th>submitter</th>\n",
       "      <th>ftp_path</th>\n",
       "      <th>img_id</th>\n",
       "      <th>...</th>\n",
       "      <th>coding_density</th>\n",
       "      <th>completeness</th>\n",
       "      <th>contamination</th>\n",
       "      <th>strain_heterogeneity</th>\n",
       "      <th>markers</th>\n",
       "      <th>5s_rrna</th>\n",
       "      <th>16s_rrna</th>\n",
       "      <th>23s_rrna</th>\n",
       "      <th>trnas</th>\n",
       "      <th>draft_quality</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>#genome</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <td>G000005825</td>\n",
       "      <td>GCF_000005825.2</td>\n",
       "      <td>ASM582v2</td>\n",
       "      <td>GCF_000005825.2</td>\n",
       "      <td>PRJNA224116</td>\n",
       "      <td>SAMN02603086</td>\n",
       "      <td>NaN</td>\n",
       "      <td>2010/12/15</td>\n",
       "      <td>Center for Genomic Sciences, Allegheny-Singer ...</td>\n",
       "      <td>ftp://ftp.ncbi.nlm.nih.gov/genomes/all/GCF/000...</td>\n",
       "      <td>646311908</td>\n",
       "      <td>...</td>\n",
       "      <td>85.144124</td>\n",
       "      <td>98.68</td>\n",
       "      <td>1.32</td>\n",
       "      <td>0.0</td>\n",
       "      <td>377</td>\n",
       "      <td>yes</td>\n",
       "      <td>yes</td>\n",
       "      <td>yes</td>\n",
       "      <td>20</td>\n",
       "      <td>high</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>G000006175</td>\n",
       "      <td>GCF_000006175.1</td>\n",
       "      <td>ASM617v2</td>\n",
       "      <td>GCF_000006175.1</td>\n",
       "      <td>PRJNA224116</td>\n",
       "      <td>SAMN00000040</td>\n",
       "      <td>NaN</td>\n",
       "      <td>2010/06/03</td>\n",
       "      <td>US DOE Joint Genome Institute (JGI-PGF)</td>\n",
       "      <td>ftp://ftp.ncbi.nlm.nih.gov/genomes/all/GCF/000...</td>\n",
       "      <td>646564549</td>\n",
       "      <td>...</td>\n",
       "      <td>80.167033</td>\n",
       "      <td>99.05</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.0</td>\n",
       "      <td>165</td>\n",
       "      <td>no</td>\n",
       "      <td>yes</td>\n",
       "      <td>yes</td>\n",
       "      <td>19</td>\n",
       "      <td>medium</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>G000006605</td>\n",
       "      <td>GCF_000006605.1</td>\n",
       "      <td>ASM660v1</td>\n",
       "      <td>GCF_000006605.1</td>\n",
       "      <td>PRJNA224116</td>\n",
       "      <td>SAMEA3283089</td>\n",
       "      <td>NaN</td>\n",
       "      <td>2005/06/27</td>\n",
       "      <td>Bielefeld Univ</td>\n",
       "      <td>ftp://ftp.ncbi.nlm.nih.gov/genomes/all/GCF/000...</td>\n",
       "      <td>637000085</td>\n",
       "      <td>...</td>\n",
       "      <td>89.378688</td>\n",
       "      <td>100.00</td>\n",
       "      <td>0.68</td>\n",
       "      <td>0.0</td>\n",
       "      <td>319</td>\n",
       "      <td>yes</td>\n",
       "      <td>yes</td>\n",
       "      <td>yes</td>\n",
       "      <td>20</td>\n",
       "      <td>high</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>G000006725</td>\n",
       "      <td>GCF_000006725.1</td>\n",
       "      <td>ASM672v1</td>\n",
       "      <td>GCF_000006725.1</td>\n",
       "      <td>PRJNA224116</td>\n",
       "      <td>SAMN02603773</td>\n",
       "      <td>NaN</td>\n",
       "      <td>2004/06/04</td>\n",
       "      <td>Sao Paulo state (Brazil) Consortium</td>\n",
       "      <td>ftp://ftp.ncbi.nlm.nih.gov/genomes/all/GCF/000...</td>\n",
       "      <td>637000348</td>\n",
       "      <td>...</td>\n",
       "      <td>82.592990</td>\n",
       "      <td>99.59</td>\n",
       "      <td>0.18</td>\n",
       "      <td>0.0</td>\n",
       "      <td>325</td>\n",
       "      <td>yes</td>\n",
       "      <td>yes</td>\n",
       "      <td>yes</td>\n",
       "      <td>20</td>\n",
       "      <td>high</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <td>G000006745</td>\n",
       "      <td>GCF_000006745.1</td>\n",
       "      <td>ASM674v1</td>\n",
       "      <td>GCF_000006745.1</td>\n",
       "      <td>PRJNA57623</td>\n",
       "      <td>SAMN02603969</td>\n",
       "      <td>NaN</td>\n",
       "      <td>2001/01/09</td>\n",
       "      <td>TIGR</td>\n",
       "      <td>ftp://ftp.ncbi.nlm.nih.gov/genomes/all/GCF/000...</td>\n",
       "      <td>637000333</td>\n",
       "      <td>...</td>\n",
       "      <td>86.533164</td>\n",
       "      <td>99.86</td>\n",
       "      <td>0.03</td>\n",
       "      <td>0.0</td>\n",
       "      <td>360</td>\n",
       "      <td>yes</td>\n",
       "      <td>yes</td>\n",
       "      <td>yes</td>\n",
       "      <td>20</td>\n",
       "      <td>high</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows × 53 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "           # assembly_accession  asm_name assembly_accession   bioproject  \\\n",
       "#genome                                                                     \n",
       "G000005825      GCF_000005825.2  ASM582v2    GCF_000005825.2  PRJNA224116   \n",
       "G000006175      GCF_000006175.1  ASM617v2    GCF_000006175.1  PRJNA224116   \n",
       "G000006605      GCF_000006605.1  ASM660v1    GCF_000006605.1  PRJNA224116   \n",
       "G000006725      GCF_000006725.1  ASM672v1    GCF_000006725.1  PRJNA224116   \n",
       "G000006745      GCF_000006745.1  ASM674v1    GCF_000006745.1   PRJNA57623   \n",
       "\n",
       "               biosample wgs_master seq_rel_date  \\\n",
       "#genome                                            \n",
       "G000005825  SAMN02603086        NaN   2010/12/15   \n",
       "G000006175  SAMN00000040        NaN   2010/06/03   \n",
       "G000006605  SAMEA3283089        NaN   2005/06/27   \n",
       "G000006725  SAMN02603773        NaN   2004/06/04   \n",
       "G000006745  SAMN02603969        NaN   2001/01/09   \n",
       "\n",
       "                                                    submitter  \\\n",
       "#genome                                                         \n",
       "G000005825  Center for Genomic Sciences, Allegheny-Singer ...   \n",
       "G000006175            US DOE Joint Genome Institute (JGI-PGF)   \n",
       "G000006605                                     Bielefeld Univ   \n",
       "G000006725                Sao Paulo state (Brazil) Consortium   \n",
       "G000006745                                               TIGR   \n",
       "\n",
       "                                                     ftp_path     img_id  ...  \\\n",
       "#genome                                                                   ...   \n",
       "G000005825  ftp://ftp.ncbi.nlm.nih.gov/genomes/all/GCF/000...  646311908  ...   \n",
       "G000006175  ftp://ftp.ncbi.nlm.nih.gov/genomes/all/GCF/000...  646564549  ...   \n",
       "G000006605  ftp://ftp.ncbi.nlm.nih.gov/genomes/all/GCF/000...  637000085  ...   \n",
       "G000006725  ftp://ftp.ncbi.nlm.nih.gov/genomes/all/GCF/000...  637000348  ...   \n",
       "G000006745  ftp://ftp.ncbi.nlm.nih.gov/genomes/all/GCF/000...  637000333  ...   \n",
       "\n",
       "           coding_density completeness contamination strain_heterogeneity  \\\n",
       "#genome                                                                     \n",
       "G000005825      85.144124        98.68          1.32                  0.0   \n",
       "G000006175      80.167033        99.05          0.00                  0.0   \n",
       "G000006605      89.378688       100.00          0.68                  0.0   \n",
       "G000006725      82.592990        99.59          0.18                  0.0   \n",
       "G000006745      86.533164        99.86          0.03                  0.0   \n",
       "\n",
       "           markers 5s_rrna  16s_rrna  23s_rrna trnas draft_quality  \n",
       "#genome                                                             \n",
       "G000005825     377     yes       yes       yes    20          high  \n",
       "G000006175     165      no       yes       yes    19        medium  \n",
       "G000006605     319     yes       yes       yes    20          high  \n",
       "G000006725     325     yes       yes       yes    20          high  \n",
       "G000006745     360     yes       yes       yes    20          high  \n",
       "\n",
       "[5 rows x 53 columns]"
      ]
     },
     "execution_count": 23,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "dfg.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "metadata": {},
   "outputs": [],
   "source": [
    "dfg_c = dfg.copy()\n",
    "dfg_c = dfg_c.set_index('# assembly_accession')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "metadata": {},
   "outputs": [],
   "source": [
    "dfg_c.to_csv('data/wol_supplemental_metadata.tsv', sep='\\t')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "metadata": {},
   "outputs": [],
   "source": [
    "tree_tips = [node.name for node in tree.tips()]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "na                       5567\n",
       "representative genome    4904\n",
       "reference genome          104\n",
       "Name: refseq_category, dtype: int64"
      ]
     },
     "execution_count": 27,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "dfg['refseq_category'].value_counts()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 28,
   "metadata": {},
   "outputs": [],
   "source": [
    "rep_ref_genomes = dfg.loc[dfg['refseq_category'].isin(['representative genome', 'reference genome'])].index"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 29,
   "metadata": {},
   "outputs": [],
   "source": [
    "def calc_brlen_metrics(tree):\n",
    "    \"\"\"Calculate branch length-related metrics.\n",
    "    \n",
    "    Originally from Zhu et al.\n",
    "\n",
    "    Parameters\n",
    "    ----------\n",
    "    tree : skbio.TreeNode\n",
    "\n",
    "    Notes\n",
    "    -----\n",
    "    The following metrics are calculated:\n",
    "    \n",
    "        - height: Sum of branch lengths from the root to the node.\n",
    "\n",
    "        - depths: Sums of branch lengths from all descendants to current node.\n",
    "\n",
    "        - red: Relative evolutionary divergence (RED), introduced by Parks,\n",
    "          et al., 2018, Nat Biotechnol.\n",
    "\n",
    "              RED = p + (d / u) * (1 - p)\n",
    "\n",
    "          where p = RED of parent, d = length, u = mean depth of parent\n",
    "    \n",
    "    Metrics will be appended to each node of tree in place.\n",
    "    \"\"\"\n",
    "    # calculate depths\n",
    "    for node in tree.postorder(include_self=True):\n",
    "        if node.name is None:\n",
    "            raise ValueError('Error: Found an unnamed node.')\n",
    "        if node.length is None:\n",
    "            node.length = 0.0\n",
    "        if node.is_tip():\n",
    "            node.depths = [0.0]\n",
    "            node.taxa = [node.name]\n",
    "        else:\n",
    "            node.depths = [\n",
    "                y + x.length for x in node.children for y in x.depths]\n",
    "            node.taxa = sorted(set().union(*[x.taxa for x in node.children]))\n",
    "\n",
    "    # calculate heights and REDs\n",
    "    for node in tree.preorder(include_self=True):\n",
    "        if node.is_root():\n",
    "            node.height = 0.0\n",
    "            node.red = 0.0\n",
    "        else:\n",
    "            node.height = node.parent.height + node.length\n",
    "            if node.is_tip():\n",
    "                node.red = 1.0\n",
    "            else:\n",
    "                node.red = node.parent.red + node.length \\\n",
    "                    / (node.length + sum(node.depths) / len(node.depths)) \\\n",
    "                    * (1 - node.parent.red)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 30,
   "metadata": {},
   "outputs": [],
   "source": [
    "calc_brlen_metrics(tree)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 259,
   "metadata": {},
   "outputs": [],
   "source": [
    "# value_counts = [dfg.loc[node.taxa]['lv2_group'].value_counts()\n",
    "#                 for node in groupings[0]]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 260,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "1917"
      ]
     },
     "execution_count": 260,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "len(groupings)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Sample genomes from within clade by WoL criteria:\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Criterium 1: Most marker genes."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 261,
   "metadata": {},
   "outputs": [],
   "source": [
    "g2markers = dfg['markers'].to_dict()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 262,
   "metadata": {},
   "outputs": [],
   "source": [
    "def most_markers(gs, g2markers):\n",
    "    max_gs = []\n",
    "    max_markers = 0\n",
    "    for g in sorted(gs, key=lambda x: g2markers[x], reverse=True):\n",
    "        if max_markers == 0:\n",
    "            max_markers = g2markers[g]\n",
    "            max_gs.append(g)\n",
    "        elif max_markers == g2markers[g]:\n",
    "            max_gs.append(g)\n",
    "        else:\n",
    "            break\n",
    "        max_gs.append(g)\n",
    "    return max_gs, max_markers"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Criterium 2: Lowest contamination."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 263,
   "metadata": {},
   "outputs": [],
   "source": [
    "g2contam = dfg['contamination'].to_dict()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 264,
   "metadata": {},
   "outputs": [],
   "source": [
    "def least_contaminated(gs, g2contam):\n",
    "    min_gs = []\n",
    "    min_contam = None\n",
    "    for g in sorted(gs, key=lambda x: g2contam[x]):\n",
    "        if min_contam is None:\n",
    "            min_contam = g2contam[g]\n",
    "            min_gs.append(g)\n",
    "        elif min_contam == g2contam[g]:\n",
    "            min_gs.append(g)\n",
    "        else:\n",
    "            break\n",
    "    return min_gs, min_contam"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Criterium 3: Highest DNA quality score."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 265,
   "metadata": {},
   "outputs": [],
   "source": [
    "g2dnaqty = dfg['score_fna'].to_dict()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 266,
   "metadata": {},
   "outputs": [],
   "source": [
    "def best_dna(gs, g2dnaqty):\n",
    "    max_gs = []\n",
    "    max_dnaqty = 0\n",
    "    for g in sorted(gs, key=lambda x: g2dnaqty[x], reverse=True):\n",
    "        if max_dnaqty == 0:\n",
    "            max_dnaqty = g2dnaqty[g]\n",
    "            max_gs.append(g)\n",
    "        elif max_dnaqty == g2dnaqty[g]:\n",
    "            max_gs.append(g)\n",
    "        else:\n",
    "            break\n",
    "    return max_gs, max_dnaqty"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 267,
   "metadata": {},
   "outputs": [],
   "source": [
    "g2species = dfg['species'].to_dict()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 268,
   "metadata": {},
   "outputs": [],
   "source": [
    "g2refseq_category = dfg['refseq_category'].to_dict()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 269,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "na                       5567\n",
       "representative genome    4904\n",
       "reference genome          104\n",
       "Name: refseq_category, dtype: int64"
      ]
     },
     "execution_count": 269,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "dfg['refseq_category'].value_counts()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 270,
   "metadata": {},
   "outputs": [],
   "source": [
    "categories = ['representative genome', 'reference genome']\n",
    "def is_complete_genome(gs, g2refseq_category):\n",
    "    new_gs = g2refseq_category.loc[gs]\n",
    "    new_gs = new_gs.loc[new_gs['refseq_category'].isin(categories)]\n",
    "    new_gs = list(new_gs.index)\n",
    "    if len(new_gs) > 0:\n",
    "        return new_gs, True\n",
    "    else:\n",
    "        return gs, False\n",
    "        "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 284,
   "metadata": {},
   "outputs": [],
   "source": [
    "from itertools import product"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 295,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "min-red: 0.5\tn_clades: 5\tn_groups: 443\n",
      "min-red: 0.5\tn_clades: 10\tn_groups: 443\n",
      "min-red: 0.5\tn_clades: 15\tn_groups: 443\n",
      "min-red: 0.5\tn_clades: 20\tn_groups: 443\n",
      "min-red: 0.75\tn_clades: 5\tn_groups: 1917\n",
      "min-red: 0.75\tn_clades: 10\tn_groups: 1917\n",
      "min-red: 0.75\tn_clades: 15\tn_groups: 1917\n",
      "min-red: 0.75\tn_clades: 20\tn_groups: 1917\n",
      "min-red: 0.9\tn_clades: 5\tn_groups: 4700\n",
      "min-red: 0.9\tn_clades: 10\tn_groups: 4700\n",
      "min-red: 0.9\tn_clades: 15\tn_groups: 4700\n",
      "min-red: 0.9\tn_clades: 20\tn_groups: 4700\n"
     ]
    }
   ],
   "source": [
    "min_reds = [0.5, 0.75, 0.9] # between 0 and 1\n",
    "n_clades_s = [5, 10, 15, 20]\n",
    "save = True\n",
    "reads_per_class = 10 ** 4\n",
    "output_dir = 'data/subsets'\n",
    "\n",
    "## start loop\n",
    "for min_red, n_clades in product(min_reds, n_clades_s):\n",
    "    clade_roots = []\n",
    "    candidates = [tree]\n",
    "    while candidates:\n",
    "        node = candidates.pop()\n",
    "        if node.red >= min_red:\n",
    "            clade_roots.append(node)\n",
    "        else:\n",
    "            candidates.extend(child for child in node)\n",
    "\n",
    "    tree_dict = {node.name: node for node in tree.traverse()}\n",
    "\n",
    "    groupings = []\n",
    "    for clade_root in clade_roots:\n",
    "\n",
    "        if len(list(clade_root.tips())) < n_clades:\n",
    "            groupings.append([])\n",
    "            continue\n",
    "\n",
    "        chosen_clades = [(sum(node.red for node in clade_root),\n",
    "                         clade_root.name)]\n",
    "\n",
    "        while len(chosen_clades) < n_clades:\n",
    "            # if chosen_clades[0][0] < 3:\n",
    "            red_score, node_name = heapq.heappop(chosen_clades)\n",
    "            node = tree_dict[node_name] \n",
    "            if node.children:\n",
    "                for child in node:\n",
    "                    heapq.heappush(chosen_clades,\n",
    "                                   (sum(gc.red for gc in child),\n",
    "                                    child.name))\n",
    "            else:\n",
    "                heapq.heappush(chosen_clades, (3, node.name))\n",
    "\n",
    "        groupings.append([tree_dict[node_name] for \n",
    "                          _, node_name in chosen_clades])\n",
    "\n",
    "\n",
    "    print(\"min-red: {}\\tn_clades: {}\\tn_groups: {}\".format(min_red, n_clades, len(groupings)))\n",
    "\n",
    "    genomes_groups = []\n",
    "    for i, group in enumerate(groupings):\n",
    "        # print(\"\\nGroup {}: n_descendants {}\".format(i, sum(len(list(node.tips())) for node in group)))\n",
    "        chosen_clades = [node.name for node in group]\n",
    "        chosen_genomes = dict()\n",
    "        for nid in chosen_clades:\n",
    "            gs = tree_dict[nid].taxa\n",
    "\n",
    "            gs, complete = is_complete_genome(gs, dfg)\n",
    "            if not complete:\n",
    "                # print(\"Warning: {} has no complete child genomes\".format(nid))\n",
    "                continue\n",
    "            # maximize marker count\n",
    "\n",
    "            gs, max_markers = most_markers(gs, g2markers)\n",
    "\n",
    "            if len(gs) > 1:\n",
    "                # minimize contamination\n",
    "                gs, min_contam = least_contaminated(gs, g2contam)\n",
    "\n",
    "            if len(gs) > 1:\n",
    "                # maximize DNA quality\n",
    "                gs, max_dnaqty = best_dna(gs, g2dnaqty)\n",
    "\n",
    "            if len(gs) > 1:\n",
    "                # random choice (UPDATE: choose first one for stability)\n",
    "                # print('Clade %s: Equally good: %s.' % (nid, ', '.join(gs)))\n",
    "                g = gs[0] # choice(gs)\n",
    "\n",
    "            else:\n",
    "                g = max(gs)\n",
    "\n",
    "            chosen_genomes[nid] = g\n",
    "\n",
    "        if len(chosen_genomes) < n_clades:\n",
    "            continue\n",
    "        # print(\"\\nGroup {}: n_descendants {}\".format(i, sum(len(list(node.tips())) for node in group)))\n",
    "\n",
    "        genomes_groups.append(chosen_genomes.items())\n",
    "        add_rows = []\n",
    "        for nid, g in sorted(chosen_genomes.items(), key=lambda x: g2species[x[1]]):\n",
    "            add_rows.append([nid, g, reads_per_class, g2species[g]])\n",
    "            # print('%s - %s: %s' % (nid, g, g2species[g]))\n",
    "\n",
    "        # get a version of tree sheared to \n",
    "        #  a) the tree only containg selected genomes\n",
    "        if save:\n",
    "            try:\n",
    "                sheared_minimal = tree.shear(list(chosen_genomes.values()))\n",
    "                tree_name = 'min-red_{}__n_{}__group_{}__minimal-tree.nwk'.format(min_red, n_clades, i)\n",
    "                # # can run the following if you want tips to \n",
    "                # invert_nodes = {val: key for key, val in chosen_genomes.items()}\n",
    "                # for tip in sheared_minimal.tips():\n",
    "                #     tip.name = invert_nodes[tip.name]\n",
    "                out_filename = os.path.join(output_dir, tree_name)\n",
    "                sheared_minimal.write(out_filename)\n",
    "                #  b) the tree that has all nodes descending from selected clades\n",
    "                all_descendants = [tree_dict[node].tips() for node in chosen_genomes.keys()]\n",
    "                sheared_all_descendants = tree.shear([item.name for sublist in all_descendants for item in sublist])\n",
    "                tree_name = 'min-red_{}__n_{}__group_{}__all-desc-tree.nwk'.format(min_red, n_clades, i)\n",
    "                out_filename = os.path.join(output_dir, tree_name)\n",
    "                sheared_all_descendants.write(out_filename)\n",
    "\n",
    "\n",
    "                df = pd.DataFrame(add_rows, columns=['class', '#genome', 'n_reads', 'species_name'])\n",
    "                filtered_dfg = dfg.loc[dfg.index.isin(df['#genome'])].to_dict()\n",
    "                accessions = [filtered_dfg['assembly_accession'][genome] for genome in df['#genome']]\n",
    "                # output_dir = 'data'\n",
    "                df['id'] = accessions\n",
    "                df.drop(['#genome'], axis=1, inplace=True)\n",
    "                f_name = 'min-red_{}__n_{}__group_{}__genome-ids.txt'.format(min_red, n_clades, i)\n",
    "                out_filename = os.path.join(output_dir, f_name)\n",
    "                df.to_csv(out_filename, index=None, sep='\\t')\n",
    "            except:\n",
    "                pass"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 310,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "min-red: 0.5\tn_clades: 5\n",
      "Number of nodes: 1605.\n",
      "min-red: 0.5\tn_clades: 10\n",
      "Number of nodes: 1603.\n",
      "min-red: 0.5\tn_clades: 15\n",
      "Number of nodes: 1603.\n",
      "min-red: 0.5\tn_clades: 20\n",
      "Number of nodes: 1603.\n",
      "min-red: 0.75\tn_clades: 5\n",
      "Number of nodes: 389.\n",
      "min-red: 0.75\tn_clades: 10\n",
      "Number of nodes: 387.\n",
      "min-red: 0.75\tn_clades: 15\n",
      "Number of nodes: 387.\n",
      "min-red: 0.75\tn_clades: 20\n",
      "Number of nodes: 387.\n",
      "min-red: 0.9\tn_clades: 5\n",
      "Number of nodes: 73.\n",
      "min-red: 0.9\tn_clades: 10\n",
      "Number of nodes: 73.\n",
      "min-red: 0.9\tn_clades: 15\n",
      "Number of nodes: 69.\n",
      "min-red: 0.9\tn_clades: 20\n"
     ]
    },
    {
     "ename": "FileNotFoundError",
     "evalue": "[Errno 2] No such file or directory: 'data/subsets/min-red_0.9__n_20__group_0__all-desc-tree.nwk'",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mFileNotFoundError\u001b[0m                         Traceback (most recent call last)",
      "\u001b[0;32m~/Documents/repos/taxonomy-assignment/taxa-assign-benchmarking/notebooks/tree_to_taxonomy.py\u001b[0m in \u001b[0;36m<module>\u001b[0;34m\u001b[0m\n\u001b[1;32m     36\u001b[0m     \u001b[0mos\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mmakedirs\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mout_dir\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m     37\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m---> 38\u001b[0;31m \u001b[0mtree\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mTreeNode\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mread\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0margs\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mtree\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m     39\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m     40\u001b[0m \u001b[0mprint\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m'Number of nodes: %d.'\u001b[0m \u001b[0;34m%\u001b[0m \u001b[0mtree\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mcount\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m~/miniconda3/envs/phylorank-test/lib/python3.6/site-packages/skbio/io/registry.py\u001b[0m in \u001b[0;36mread\u001b[0;34m(cls, file, format, **kwargs)\u001b[0m\n\u001b[1;32m    650\u001b[0m         \u001b[0;34m@\u001b[0m\u001b[0mclassonlymethod\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    651\u001b[0m         \u001b[0;32mdef\u001b[0m \u001b[0mread\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mcls\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mfile\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mformat\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0;32mNone\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m**\u001b[0m\u001b[0mkwargs\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m--> 652\u001b[0;31m             \u001b[0;32mreturn\u001b[0m \u001b[0mregistry\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mread\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mfile\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0minto\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0mcls\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mformat\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0mformat\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m**\u001b[0m\u001b[0mkwargs\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m    653\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    654\u001b[0m         \u001b[0mimports\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mregistry\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0m_import_paths\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mread_formats\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m~/miniconda3/envs/phylorank-test/lib/python3.6/site-packages/skbio/io/registry.py\u001b[0m in \u001b[0;36mread\u001b[0;34m(self, file, format, into, verify, **kwargs)\u001b[0m\n\u001b[1;32m    511\u001b[0m                 \u001b[0;32mreturn\u001b[0m \u001b[0;34m(\u001b[0m\u001b[0mx\u001b[0m \u001b[0;32mfor\u001b[0m \u001b[0mx\u001b[0m \u001b[0;32min\u001b[0m \u001b[0;34m[\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    512\u001b[0m         \u001b[0;32melse\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m--> 513\u001b[0;31m             \u001b[0;32mreturn\u001b[0m \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0m_read_ret\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mfile\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mformat\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0minto\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mverify\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mkwargs\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m    514\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    515\u001b[0m     \u001b[0;32mdef\u001b[0m \u001b[0m_read_ret\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mself\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mfile\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mfmt\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0minto\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mverify\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mkwargs\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m~/miniconda3/envs/phylorank-test/lib/python3.6/site-packages/skbio/io/registry.py\u001b[0m in \u001b[0;36m_read_ret\u001b[0;34m(self, file, fmt, into, verify, kwargs)\u001b[0m\n\u001b[1;32m    515\u001b[0m     \u001b[0;32mdef\u001b[0m \u001b[0m_read_ret\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mself\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mfile\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mfmt\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0minto\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mverify\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mkwargs\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    516\u001b[0m         \u001b[0mio_kwargs\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0m_find_io_kwargs\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mkwargs\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m--> 517\u001b[0;31m         \u001b[0;32mwith\u001b[0m \u001b[0m_resolve_file\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mfile\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m**\u001b[0m\u001b[0mio_kwargs\u001b[0m\u001b[0;34m)\u001b[0m \u001b[0;32mas\u001b[0m \u001b[0;34m(\u001b[0m\u001b[0mfile\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0m_\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0m_\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m    518\u001b[0m             reader, kwargs = self._init_reader(file, fmt, into, verify, kwargs,\n\u001b[1;32m    519\u001b[0m                                                io_kwargs)\n",
      "\u001b[0;32m~/miniconda3/envs/phylorank-test/lib/python3.6/contextlib.py\u001b[0m in \u001b[0;36m__enter__\u001b[0;34m(self)\u001b[0m\n\u001b[1;32m     79\u001b[0m     \u001b[0;32mdef\u001b[0m \u001b[0m__enter__\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mself\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m     80\u001b[0m         \u001b[0;32mtry\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m---> 81\u001b[0;31m             \u001b[0;32mreturn\u001b[0m \u001b[0mnext\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mgen\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m     82\u001b[0m         \u001b[0;32mexcept\u001b[0m \u001b[0mStopIteration\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m     83\u001b[0m             \u001b[0;32mraise\u001b[0m \u001b[0mRuntimeError\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m\"generator didn't yield\"\u001b[0m\u001b[0;34m)\u001b[0m \u001b[0;32mfrom\u001b[0m \u001b[0;32mNone\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m~/miniconda3/envs/phylorank-test/lib/python3.6/site-packages/skbio/io/util.py\u001b[0m in \u001b[0;36m_resolve_file\u001b[0;34m(file, **kwargs)\u001b[0m\n\u001b[1;32m    201\u001b[0m \u001b[0;34m@\u001b[0m\u001b[0mcontextmanager\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    202\u001b[0m \u001b[0;32mdef\u001b[0m \u001b[0m_resolve_file\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mfile\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m**\u001b[0m\u001b[0mkwargs\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m--> 203\u001b[0;31m     \u001b[0mfile\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0msource\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mis_binary_file\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0m_resolve\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mfile\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m**\u001b[0m\u001b[0mkwargs\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m    204\u001b[0m     \u001b[0;32mtry\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    205\u001b[0m         \u001b[0;32myield\u001b[0m \u001b[0mfile\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0msource\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mis_binary_file\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m~/miniconda3/envs/phylorank-test/lib/python3.6/site-packages/skbio/io/util.py\u001b[0m in \u001b[0;36m_resolve\u001b[0;34m(file, mode, encoding, errors, newline, compression, compresslevel)\u001b[0m\n\u001b[1;32m     55\u001b[0m         \u001b[0msource\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0msource_handler\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mfile\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0marguments\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m     56\u001b[0m         \u001b[0;32mif\u001b[0m \u001b[0mmode\u001b[0m \u001b[0;34m==\u001b[0m \u001b[0;34m'r'\u001b[0m \u001b[0;32mand\u001b[0m \u001b[0msource\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mcan_read\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m---> 57\u001b[0;31m             \u001b[0mnewfile\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0msource\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mget_reader\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m     58\u001b[0m             \u001b[0;32mbreak\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m     59\u001b[0m         \u001b[0;32melif\u001b[0m \u001b[0mmode\u001b[0m \u001b[0;34m==\u001b[0m \u001b[0;34m'w'\u001b[0m \u001b[0;32mand\u001b[0m \u001b[0msource\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mcan_write\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m~/miniconda3/envs/phylorank-test/lib/python3.6/site-packages/skbio/io/_iosources.py\u001b[0m in \u001b[0;36mget_reader\u001b[0;34m(self)\u001b[0m\n\u001b[1;32m     92\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m     93\u001b[0m     \u001b[0;32mdef\u001b[0m \u001b[0mget_reader\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mself\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m---> 94\u001b[0;31m         \u001b[0;32mreturn\u001b[0m \u001b[0mio\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mopen\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mfile\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mmode\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0;34m'rb'\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m     95\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m     96\u001b[0m     \u001b[0;32mdef\u001b[0m \u001b[0mget_writer\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mself\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;31mFileNotFoundError\u001b[0m: [Errno 2] No such file or directory: 'data/subsets/min-red_0.9__n_20__group_0__all-desc-tree.nwk'"
     ]
    }
   ],
   "source": [
    "for min_red, n_clades in product(min_reds, n_clades_s):\n",
    "    print(\"min-red: {}\\tn_clades: {}\".format(min_red, n_clades))\n",
    "    in_tree = \"data/subsets/min-red_{}__n_{}__group_0__all-desc-tree.nwk\".format(min_red, n_clades)\n",
    "    out_dir = \"taxdumps/min-red_{}__n_{}__group_0_taxdump\".format(min_red, n_clades)\n",
    "    %run -i ./tree_to_taxonomy.py $in_tree --out-directory $out_dir\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 299,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Number of nodes: 1605.\n"
     ]
    }
   ],
   "source": [
    "%run -i ./tree_to_taxonomy.py  --out-directory "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python (skbio-1)",
   "language": "python",
   "name": "skbio"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.6.9"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
